vllm/v1/sample/ops/topk_topp_sampler.py

```diff
@@ -138,8 +138,23 @@ def apply_top_k_top_p(
 
     This function sorts the logits tensor, which can be slow for large batches.
     """
-    if k is None and p is None:
+    if p is None:
+        if k is None:
+            return logits
+
+        # Avoid sorting vocab for top-k only case.
+
+        no_top_k_mask = k == logits.shape[1]
+        # Set non-top-k rows to 1 so that we can gather.
+        k = k.masked_fill(no_top_k_mask, 1).unsqueeze(1)
+        max_top_k = k.max()
+        k.sub_(1)
```

Comment:
A dumb question: Why -1 here?

Reply:
Your questions are never dumb @WoosukKwon :)

It's just converting to 0-based index for the gather. `top_k_mask.shape[1]` will be `max_top_k`, so the request with max top k will be out-of-bounds for example.

And we're masking logits that are less than these selected ones so I think it works out (e.g. for k=1 we want to select the 0th element of second dimension and then exclude all with lower logit value than that...)

Reply:
Got it. Thanks for the explanation.
Could you also put it in the code so that anyone can easily understand?